File: AdventOfCode2024/Day1/day_1.rb
class Day1
  INPUT_PATH = 'Day1/input.txt'.freeze

  def self.puzzle_1
    # Read and process input in a single pass
    locations = File.readlines(INPUT_PATH, chomp: true).map do |line|
      left, right = line.split
      [left.to_i, right.to_i]
    end.transpose

    # Calculate total distance using array operations
    left_sorted, right_sorted = locations.map(&:sort)
    left_sorted.zip(right_sorted).sum { |left, right| (left - right).abs }
  end

  def self.puzzle_2
    # Read file once and split into two arrays
    locations = File.readlines(INPUT_PATH, chomp: true).map do |line|
      left, right = line.split
      [left.to_i, right.to_i]
    end.transpose

    left_locations, right_locations = locations

    # Create hash map counting occurrences in right locations
    right_location_map = Hash.new(0)
    right_locations.each { |num| right_location_map[num] += 1 }

    similarity_score = 0
    left_locations.each do |left_location|
      similarity_score += (left_location * right_location_map[left_location])
    end

    similarity_score
  end
end

puts Day1.puzzle_1
puts Day1.puzzle_2

File: AdventOfCode2024/java/src/main/java/day1/Day1.java
package day1;

import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import java.io.IOException;
import java.util.stream.IntStream;
import java.util.ArrayList;
import java.util.Scanner;
import java.util.HashMap;
import java.util.Map;

public class Day1 {
    public static final String INPUT_PATH = "../../../../inputs/day1/input.txt";

    public static void main(String[] args) {
        System.out.println(puzzle1());
        System.out.println(puzzle2());
    }

    public static int puzzle1() {
        int totalDistance = 0;
        
        List<Integer> leftLocations = new ArrayList<>();
        List<Integer> rightLocations = new ArrayList<>();
        
        try (Scanner scanner = new Scanner(Paths.get(INPUT_PATH))) {
            while (scanner.hasNextLine()) {
                String line = scanner.nextLine().trim();
                if (!line.isEmpty()) {
                    String[] parts = line.split("\\s+");
                    leftLocations.add(Integer.parseInt(parts[0]));
                    rightLocations.add(Integer.parseInt(parts[1]));
                }
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
        
        leftLocations.sort(Integer::compareTo);
        rightLocations.sort(Integer::compareTo);

        totalDistance = IntStream.range(0, leftLocations.size()).map(
            i -> Math.abs(leftLocations.get(i) - rightLocations.get(i))
        ).sum();
        return totalDistance;
    }

    public static int puzzle2() {
         List<Integer> leftLocations = new ArrayList<>();
        List<Integer> rightLocations = new ArrayList<>();
        
        try (Scanner scanner = new Scanner(Paths.get(INPUT_PATH))) {
            while (scanner.hasNextLine()) {
                String line = scanner.nextLine().trim();
                if (!line.isEmpty()) {
                    String[] parts = line.split("\\s+");
                    leftLocations.add(Integer.parseInt(parts[0]));
                    rightLocations.add(Integer.parseInt(parts[1]));
                }
            }
        } catch (IOException e) {
            e.printStackTrace();
        }

        Map<Integer, Integer> rightLocationsMap = new HashMap<>();
        rightLocations.forEach(location -> rightLocationsMap.put(location, rightLocationsMap.getOrDefault(location, 0) + 1));

        return leftLocations.stream()
            .mapToInt(location -> location * rightLocationsMap.getOrDefault(location, 0))
            .sum();
    }
}
